File: php/day4/part1.php
<?php
$input = fopen("day4.txt", "r");
$arr = [];

$sum = 0;

while (!feof($input)) {
    $tempStr = fgets($input);
    $tempArr = str_split($tempStr);

    $sum += substr_count($tempStr, "XMAS");
    $sum += substr_count($tempStr, "SAMX");

    $tempArr = str_split(trim($tempStr));
    array_push($arr, $tempArr);
}
countVert($arr);

$arrUpLeft = $arr;
for($i = 0; $i<count($arr); $i++) {
    $arrUpLeft[$i] = array_pad($arrUpLeft[$i], -$i-count($arr), " ");
}
countVert($arrUpLeft);

$arrUpRight = $arr;
for($i = 0; $i<count($arr); $i++) {
    $arrUpRight[count($arr)-$i-1] = array_pad($arrUpRight[count($arr)-$i-1], -$i-count($arr), " ");
}
countVert($arrUpRight);

echo $sum;
fclose($input);

function countVert($arr) {
    global $sum;

    $arrT= array_map(null, ...$arr);
    foreach ($arrT as $tempArr) {
        $tempStr = implode("", $tempArr);

        $sum += substr_count($tempStr, "XMAS");
        $sum += substr_count($tempStr, "SAMX");
    }
}

File: python/day4/part1.py
def main():
    arr = []

    count = 0
    with open("./day4.txt", "r", ) as file:
        for line in file:
            line = line.strip()
            arr.append(list(line))

            count += line.count("XMAS")
            count += line.count("SAMX")
    
    for line in transStr(arr):
        count += line.count("XMAS")
        count += line.count("SAMX")

    upLeft = []
    for i, line in enumerate(arr):
        tempLine = i*[" "]
        tempLine.extend(line)
        tempLine.extend((len(line)-i-1)*[" "])
        upLeft.append(tempLine)
    for line in transStr(upLeft):
        count += line.count("XMAS")
        count += line.count("SAMX")

    upRight = []
    for i, line in enumerate(arr):
        tempLine = (len(line)-i-1)*[" "]
        tempLine.extend(line)
        tempLine.extend(i*[" "])
        upRight.append(tempLine)
    for line in transStr(upRight):
        count += line.count("XMAS")
        count += line.count("SAMX")
    
    print(count)

def transStr(inList: list):
    return map(lambda line : "".join(line), zip(*inList))

if __name__ == "__main__":
    main()